### PR TITLE
Avoid unnamed namespaces

### PR DESCRIPTION
**Pull request by masterleinad**
This PR avoids unnamed namespaces in header files by either removing the namespace or naming it (`internal`, `*Helper` or `*Implementation`).
Do we wan't to ban these everywhere due to possible issues with symbols created for each header file individually or allow them if we think this is safe?

### CODE REVIEW

**Comment by davydden** [comment]
> Do we wan't to ban these everywhere due to possible issues with symbols created for each header file individually

But that should be ok so long as function name/arguments are not accidentally the same, right?

File: include/deal.II/base/symmetric_tensor.h [hunk 1]
```diff
@@ -2338,87 +2338,84 @@ namespace internal
 {
   namespace SymmetricTensorImplementation
   {
-    namespace
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: I would name it ‘internal’ which suggests that users shouldn’t use it

File: include/deal.II/base/symmetric_tensor.h [hunk 2]
```diff
@@ -2437,94 +2434,91 @@ namespace internal
...
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: Ditto

File: include/deal.II/base/symmetric_tensor.h [hunk 3]
```diff
@@ -3090,24 +3084,20 @@ namespace internal
     std::array<std::pair<Number, Tensor<1, 3, Number>>, 3>
     hybrid(const dealii::SymmetricTensor<2, 3, Number> &A);
 
-    namespace
```
Comment (davydden) [inline include/deal.II/base/symmetric_tensor.h]: Same

File: include/deal.II/base/symmetric_tensor.h (hunk 1 above)
Comment (jppelteret) [reply include/deal.II/base/symmetric_tensor.h]: @davydden But this is already `internal::SymmetricTensorImplementation`...

Comment (davydden) [reply include/deal.II/base/symmetric_tensor.h]: Oh, then it’s fine. Did not notice this from phone

**Comment by masterleinad** [comment]
See https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file for some examples.

**Comment by masterleinad** [comment]
https://github.com/dealii/dealii/pull/6742#issuecomment-397351546 by @bangerth:

The issue with anonymous namespaces is that they are defined as given a unique name in every translation unit. That means that if you have
``` 
  namespace { int foo () {} }
```
in a header file, then this function will show up in one object file as the mangled form of something along the lines of `abc::foo()` and in another object file as `def::foo()`. They both have internal linkage, but because they have different names, the compiler can not eliminate duplicates. So you get code bloat at the library/executable level.

But the main reason to not use anonymous namespaces is because inevitably someone thinks that it's a good idea to put a variable in there:
```
  namespace {
    int i;
  }
```
This looks innocuous, but in reality, you now have as many variables as the number of times this header file is included elsewhere. (If you mark the variable as `const` and initialize it in one `.cc` file, then all of the source files that include the header will yield a linker error. Same if you make the variable extern.)

**Comment by masterleinad** [comment]
https://github.com/dealii/dealii/pull/6742#issuecomment-397325361 by @tamiko:

@bangerth @masterleinad Well, we can make that a more global effort. Ultimately, we should compile a  list of rules what we actually want to have. For example, we have quite a number of [1]
```
namespace internal
{
  namespace SymmetricTensorImplementation
  {
    namespace
    {
      // templated function
    }
  }
}
```
floating around in `*.h` files that look a bit silly. I have no problem promoting those to `internal::SymmetricTensorImplementation`. On the other hand we should definitely keep all anonymous namespaces in `*.templates.h` files - there they mainly suppress symbol visibility (which is a good thing).

Edit: Comment: [1] It would have made more sense to reduce this to simply `namespace { ... }`, or avoid the anonymous namespace altogether. But from a code generation point of view _all_ of these variants will trigger code generation in the current compilation unit on "ODR"-use.

**Comment by masterleinad** [comment]
> They both have internal linkage, but because they have different names, the compiler can not eliminate duplicates. So you get code bloat at the library/executable level.

But the compiler might be better in optimizing when it knows that the function is not used the current compilation unit.

> But the main reason to not use anonymous namespaces is because inevitably someone thinks that  it's a good idea to put a variable in there:
>
> namespace {int i;}

Yes, this is the case we have to be careful about. The question is if this is reason enough to disallow unnamed namespaces in header files in general or not.

> On the other hand we should definitely keep all anonymous namespaces in *.templates.h files - there they mainly suppress symbol visibility (which is a good thing).

I agree.

**Comment by tamiko** [comment]
@bangerth @masterleinad I agree that the potential symbol bloat is a problem. The stuff we tend to put in anonymous namespaces tend to not get inlined. And our typical use is of the following form:
```
namespace { template <int i> void foo() {} }
template <int i> void bar() { foo<i>(); }
```
Now, duplicate symbols for bar get eliminated but the corresponding symbol for foo stays.

Example:
```C++
// main.cc
namespace { template <int i> void foo() {} }
template <int i> void bar() { foo<i>(); }
int main() { bar<1>(); }

// consumer.cc
namespace { template <int i> void foo() {} }
template <int i> void bar() { foo<i>(); }
void consumer() { bar<1>(); }
```

```shell
$ g++ consumer.cc main.cc
$ readelf -Ws a.out | grep -P "(foo|bar)" | c++filt 
32: 000000000000069d    46 FUNC    LOCAL  DEFAULT   13 void (anonymous namespace)::foo<1>()
34: 0000000000000735    46 FUNC    LOCAL  DEFAULT   13 void (anonymous namespace)::foo<1>()
42: 00000000000006cb    51 FUNC    WEAK   DEFAULT   13 void bar<1>()
```

Sometimes optimization might help:
```shell
$ g++ -O3 consumer.cc main.cc
$ readelf -Ws a.out | grep -P "(foo|bar)" | c++filt
(nothing)
```

Or garbage collecting:
```shell
$ g++ -Wl,--gc-sections -fdata-sections -ffunction-sections consumer.cc main.cc 
$ readelf -Ws a.out | grep -P "(foo|bar)" | c++filt                            
31: 000000000000069d    46 FUNC    LOCAL  DEFAULT   13 void (anonymous namespace)::foo<1>()
40: 000000000000066a    51 FUNC    WEAK   DEFAULT   13 void bar<1>()
```

I have to say, I always thought that the latter (full garbage collecting) would be default. So, yes. anonymous namespaces have to go.

**Comment by masterleinad** [comment]
Some statistics using `clang-6`:

New:
```
$ du -sh lindeal_II*
1,3G	libdeal_II.g.so.9.1.0-pre
190M	libdeal_II.so.9.1.0-pre
```
Old:
```
1,3G	libdeal_II.g.so.9.1.0-pre
188M	libdeal_II.so.9.1.0-pre
```

New:
```
$ nm libdeal_II.g.so.9.1.0-pre | wc -l
817225
$ nm libdeal_II.so.9.1.0-pre | wc -l
283282
```

Old:
```
$ nm libdeal_II.g.so.9.1.0-pre | wc -l
818036
$ nm libdeal_II.so.9.1.0-pre | wc -l
280637
```

**Comment by tamiko** [comment]
@masterleinad I guess, the real disaster comes into play when you, for example, have 20 compilation units in user code that all trigger a ODR-used code generation and symbol emission of the same function.

**Comment by bangerth** [comment]
The symbol and size growth is annoying, though only on the order of 1%.

@tamiko -- I think we agree that the anonymous namespaces should be avoided. But then you commented that "[1] It would have made more sense to reduce this to simply namespace { ... }, or avoid the anonymous namespace altogether." That's also a bad idea since it puts everything into the global namespace as far as names (though not linkage) is concerned. We'd be guaranteed to accidentally re-use names and get funny problems.

@masterleinad -- can you please rebase to get rid of the merge conflict?

**Comment by masterleinad** [comment]
Rebased. With these changes, we just have to be careful not to run into ODR problems.

**Comment by masterleinad** [comment]
A testsuite run with `-fsanitize-address` and `ASAN_OPTIONS=detect_odr_violation=2` passes.

**Comment by bangerth** [comment]
OK to merge once the tester is happy.

/run-tests

**Comment by bangerth** [comment]
The various trade-offs above basically come down to this: putting functions into an anonymous namespace in header files reduced the number of *exported* symbols that appear in the linker interface of `libdealii.so` but to code duplication and code bloat. Not putting them into anonymous namespaces allows the compiler the merge duplicate definitions, but at the cost of more functions in the linker interface. I don't know whether there is a way to get both. What happens if you mark functions in named namespaces as `static`, i.e., the old C way? In that case, they have internal linkage and do not show up in the linker interface. But is the compiler allowed to merge definitions of such functions?

**Comment by bangerth** [comment]
Aw, build failed...

**Comment by masterleinad** [comment]
> Aw, build failed...

Unity builds don't make this easier...

**Comment by masterleinad** [comment]
> What happens if you mark functions in named namespaces as static, i.e., the old C way? In that case, they have internal linkage and do not show up in the linker interface. But is the compiler allowed to merge definitions of such functions?

That might be something to keep in mind for trying at a later time.

**Pull request merged**